lib/src/native_string.ml

```diff
@@ -1,340 +1,21 @@
 
 open Sosa_pervasives
-open Printf
-open Api
-include StringLabels
 
-module F = Functors
-
-type character = char
-
-let empty = ""
-let is_empty t = (compare "" t = 0)
-
-let of_character = make 1
-let of_character_list cl =
-  let length = List.length cl in
-  let buf = String.make length '\x00' in
-  List.iteri cl ~f:(fun i c -> buf.[i] <- c);
-  buf
-
-let to_character_list s =
-  let res = ref [] in
-  for i = length s - 1 downto 0 do
-    res := s.[i] :: !res
-  done;
-  !res
-
-let get s ~index =
-  try Some (s.[index])
-  with _ -> None
-
-let set s ~index ~v =
-  if index > String.length s - 1
-  then None
-  else begin
-    let cop = String.copy s in
-    cop.[index] <- v;
-    Some cop
-  end
-let get_exn s ~index = s.[index]
-let set_exn s ~index ~v =
-  match set s ~index ~v with None -> failwith "set_exn" | Some s -> s
-
-let compare_substring (a, idxa, lena) (b, idxb, lenb) =
-  let module With_exns = struct
-    exception Return of int
-    exception Left_out of int
-    exception Right_out of int
-    let f () =
-      try
-        let shortest = min lena lenb in
-        for i = 0 to shortest - 1 do
-          let ca = try a.[idxa + i] with _ -> raise (Left_out i) in
-          let cb = try b.[idxb + i] with _ -> raise (Right_out i) in
-          let c = Char.compare  ca cb in
-          if c <> 0
-          then raise (Return c)
-          else ()
-        done;
-        (Pervasives.compare (lena : int) lenb)
-      with
-      | Return c -> c
-      | Left_out c -> (* a went out of bounds at 'c + idxa' *) -1
-      | Right_out _ -> (* b went out of bounds at 'c + idxb' *)
-        (* so, a is “longer” *) 1
-  end in
-  With_exns.f ()
-
-type s = t
-module T_length_and_compsub = struct
-  type t = s let length = length let compare_substring = compare_substring
-end
-include F.Compare_substring_strict_of_loose(T_length_and_compsub)
-include F.Make_index_of_string(T_length_and_compsub)
+module N = Functors.Make_native(struct
+  include StringLabels
+  let of_buffer = Buffer.contents
+  let empty = ""
+  let string_for_output s = s
+  end)
 
+include N
 
-let to_native_string x = String.copy x
-let of_native_string x = return (String.copy x)
+let to_native_string x = StringLabels.copy x
+let of_native_string x = return (StringLabels.copy x)
```

Comment:
Since `String.t` are not mutable, we should not need those `copy`s anymore right?